File: CodingTest/Coding_Test/BAEKJOON/1918.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        char[] exp=sc.next().toCharArray();
        StringBuilder ans=new StringBuilder();
        Deque<Character> stack = new LinkedList<>();
        for(char c:exp){
            if(Character.isLetter(c)){
                ans.append(c);
            }else if(c=='('){
                stack.addLast(c);
            }else if(c==')'){
                while(!stack.isEmpty()&&stack.getLast()!='('){
                    ans.append(stack.removeLast());
                }
                stack.removeLast();
            }else if(c=='+'||c=='-'){
                while(!stack.isEmpty()&&stack.getLast()!='('){
                    ans.append(stack.removeLast());
                }
                stack.addLast(c);
            }else if(c=='*'||c=='/'){
                while(!stack.isEmpty()&&stack.getLast()!='('&&stack.getLast()!='+'&&stack.getLast()!='-'){
                    ans.append(stack.removeLast());
                }
                stack.addLast(c);
            }
        }
        while(!stack.isEmpty()){
            ans.append(stack.removeLast());
        }
        System.out.println(ans.toString());
    }
}

File: CodingTest/Coding_Test/BAEKJOON/1918.py
from collections import deque

exp=list(input())

ans=[]
stack=deque()
for c in exp:
    if c.isalpha():
        ans.append(c)
    elif c=='(':
        stack.append(c)
    elif c==')':
        while stack and stack[-1]!='(':
            ans.append(stack.pop())
        stack.pop()
    elif c=='+' or c=='-':
        while stack and stack[-1]!='(':
            ans.append(stack.pop())
        stack.append(c)
    elif c=='*' or c=='/':
        while stack and stack[-1]!='(' and stack[-1]!='+' and stack[-1]!='-':
            ans.append(stack.pop())
        stack.append(c)
while stack:
    ans.append(stack.pop())
answer=""
for a in ans:
    answer+=a
print(answer)